nexus/src/app/sagas/region_snapshot_replacement_start.rs

```diff
@@ -0,0 +1,1128 @@
+        .fetch()
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    Ok(db_snapshot.volume_id)
+}
+
+async fn rsrss_create_fake_volume(
+    sagactx: NexusActionContext,
+) -> Result<(), ActionError> {
+    let osagactx = sagactx.user_data();
+
+    let new_volume_id = sagactx.lookup::<Uuid>("new_volume_id")?;
+
+    // Create a fake volume record for the old snapshot target. This will be
+    // deleted after snapshot replacement has finished. It can be completely
+    // blank here, it will be replaced by `volume_replace_snapshot`.
+
+    let volume_construction_request = VolumeConstructionRequest::Volume {
+        id: new_volume_id,
+        block_size: 0,
+        sub_volumes: vec![VolumeConstructionRequest::Region {
+            block_size: 0,
+            blocks_per_extent: 0,
+            extent_count: 0,
+            gen: 0,
+            opts: CrucibleOpts {
+                id: new_volume_id,
+                target: vec![],
+                lossy: false,
+                flush_timeout: None,
+                key: None,
+                cert_pem: None,
+                key_pem: None,
+                root_cert_pem: None,
+                control: None,
+                read_only: true,
+            },
+        }],
+        read_only_parent: None,
+    };
+
+    let volume_data = serde_json::to_string(&volume_construction_request)
+        .map_err(|e| {
+            ActionError::action_failed(Error::internal_error(&e.to_string()))
+        })?;
+
+    let volume = db::model::Volume::new(new_volume_id, volume_data);
+
+    osagactx
+        .datastore()
+        .volume_create(volume)
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    Ok(())
+}
+
+async fn rsrss_create_fake_volume_undo(
+    sagactx: NexusActionContext,
+) -> Result<(), anyhow::Error> {
+    let osagactx = sagactx.user_data();
+
+    // Delete the fake volume.
+
+    let new_volume_id = sagactx.lookup::<Uuid>("new_volume_id")?;
+    osagactx.datastore().volume_hard_delete(new_volume_id).await?;
+
+    Ok(())
+}
+
+#[derive(Debug)]
+struct ReplaceParams {
+    old_volume_id: Uuid,
+    old_snapshot_address: SocketAddrV6,
+    new_region_address: SocketAddrV6,
+    new_volume_id: Uuid,
+}
+
+async fn get_replace_params(
+    sagactx: &NexusActionContext,
+) -> Result<ReplaceParams, ActionError> {
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params::<Params>()?;
+
+    let new_volume_id = sagactx.lookup::<Uuid>("new_volume_id")?;
+
+    let region_snapshot = osagactx
+        .datastore()
+        .region_snapshot_get(
+            params.request.old_dataset_id,
+            params.request.old_region_id,
+            params.request.old_snapshot_id,
+        )
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    let Some(region_snapshot) = region_snapshot else {
+        return Err(ActionError::action_failed(format!(
+            "region snapshot {} {} {} deleted!",
```

Comment:
Do we know this is deleted?  Is that what a `None` means?

Reply:
We do yeah - if the corresponding region_snapshot is missing then that means the volume references went to zero, meaning the crucible resources would have been cleaned up and the record hard deleted.